File: backup/src/jclang/NativePool.java


package jclang;

import org.jetbrains.annotations.NotNull;
import jclang.structs.CXString;

import java.util.ArrayList;
import java.util.List;

/* package */ class NativePool {
    public static final NativePool I = new NativePool();

    private final List<CXString.ByValue> strings = new ArrayList<CXString.ByValue>();
    private final List<TranslationUnit> translationUnits = new ArrayList<TranslationUnit>();
    private final List<Diagnostic> diagnostics = new ArrayList<Diagnostic>();
    private final List<CXIndexAction> indexActions = new ArrayList<CXIndexAction>();
    private final List<Index> indexes = new ArrayList<Index>();

    private NativePool() {}

    public void record(@NotNull CXString.ByValue string) {
        strings.add(string);
    }

    @NotNull
    public TranslationUnit record(@NotNull TranslationUnit translationUnit) {
        translationUnits.add(translationUnit);
        return translationUnit;
    }

    @NotNull
    public Diagnostic record(@NotNull Diagnostic diagnostic) {
        diagnostics.add(diagnostic);
        return diagnostic;
    }

    @NotNull
    public CXIndexAction record(@NotNull CXIndexAction indexAction) {
        indexActions.add(indexAction);
        return indexAction;
    }

    @NotNull
    public Index record(@NotNull Index index) {
        indexes.add(index);
        return index;
    }

    public void disposeAll() {
        LibClang lib = LibClang.I;
        for (CXString.ByValue string : strings) {
            lib.disposeString(string);
        }
        for (TranslationUnit translationUnit : translationUnits) {
            lib.disposeTranslationUnit(translationUnit);
        }
        for (Diagnostic diagnostic : diagnostics) {
            lib.disposeDiagnostic(diagnostic);
        }
        for (CXIndexAction indexAction : indexActions) {
            lib.IndexAction_dispose(indexAction);
        }
        for (Index index : indexes) {
            lib.disposeIndex(index);
        }

        strings.clear();
        translationUnits.clear();
        diagnostics.clear();
        indexActions.clear();
        indexes.clear();
    }
}

File: src/main/java/klang/jvm/NativePool.kt
package klang.jvm

import klang.jvm.binding.CXIndexAction

object NativePool {
    private val translationUnits: MutableList<TranslationUnit> = mutableListOf()
    private val diagnostics: MutableList<Diagnostic> = mutableListOf()
    private val indexActions: MutableList<CXIndexAction> = mutableListOf()
    private val indexes: MutableList<Index> = mutableListOf()

    fun record(translationUnit: TranslationUnit): TranslationUnit {
        translationUnits.add(translationUnit)
        return translationUnit
    }

    fun record(diagnostic: Diagnostic): Diagnostic {
        diagnostics.add(diagnostic)
        return diagnostic
    }

    fun record(indexAction: CXIndexAction): CXIndexAction {
        indexActions.add(indexAction)
        return indexAction
    }

    fun record(index: Index): Index {
        indexes.add(index)
        return index
    }

    fun disposeAll() {
        val lib = Clang
        for (translationUnit in translationUnits) {
            lib.disposeTranslationUnit(translationUnit)
        }
        for (diagnostic in diagnostics) {
            lib.disposeDiagnostic(diagnostic)
        }
        for (indexAction in indexActions) {
            lib.IndexAction_dispose(indexAction)
        }
        for (index in indexes) {
            lib.disposeIndex(index)
        }
        translationUnits.clear()
        diagnostics.clear()
        indexActions.clear()
        indexes.clear()
    }
}